pymc/distributions/mixture.py

```diff
@@ -112,454 +132,178 @@ class Mixture(Distribution):
-                    "{}".format(self._comp_dist_shapes)
-                )
+            weights_ = weights_type
+        broadcasted_weights_ = at.broadcast_to(weights_, stacked_components_.shape)
 
-            # We wrap the _comp_dist.random by adding the kwarg raw_size_,
-            # which will be the size attribute passed to _comp_samples.
-            # _comp_samples then calls generate_samples, which may change the
-            # size value to make it compatible with scipy.stats.*.rvs
-            self._generators = []
-            for comp_dist in comp_dists:
-                generator = Mixture._comp_dist_random_wrapper(comp_dist.random)
-                self._generators.append(generator)
-            self.comp_is_distribution = False
-
-    @staticmethod
-    def _comp_dist_random_wrapper(random):
-        """Wrap the comp_dists.random method to take the kwarg raw_size_ and
-        use it's value to replace the size parameter. This is needed because
-        generate_samples makes the size value compatible with the
-        scipy.stats.*.rvs, where size has a different meaning than in the
-        distributions' random methods.
-        """
+        choices_ = at.random.categorical(broadcasted_weights_, rng=rng_type)
+        mix_out_ = at.take_along_axis(stacked_components_, choices_[..., None], axis=-1)
+        mix_out_ = at.reshape(mix_out_, components_type[0].shape)
+
+        next_rng_ = choices_.owner.outputs[0]
+        mix_op = MarginalMixtureRV(
+            inputs=[rng_type, weights_type, *components_type], outputs=[next_rng_, mix_out_]
+        )
 
-        def wrapped_random(*args, **kwargs):
-            raw_size_ = kwargs.pop("raw_size_", None)
-            # Distribution.random's signature is always (point=None, size=None)
-            # so size could be the second arg or be given as a kwarg
-            if len(args) > 1:
-                args[1] = raw_size_
+        # Create the actual MarginalMixture variable
+        next_rng, mix_out = mix_op(choices_rng, weights, *components)
+
+        # We need to set_default_updates ourselves, because the choices RV is hidden
+        # inside OpFromGraph and PyMC will never find it otherwise
+        choices_rng.default_update = next_rng
+
+        # Reference nodes to facilitate identification in other classmethods
+        mix_out.tag.weights = weights
+        mix_out.tag.components = components
+        mix_out.tag.choices_rng = choices_rng
+
+        # Component RVs terms are accounted by the Mixture logprob, so they can be
+        # safely ignore by Aeppl (this tag prevents UserWarning)
+        for component in components:
+            component.tag.ignore_logprob = True
+
+        if size is not None:
+            mix_out = cls.change_size(mix_out, size)
+
+        return mix_out
+
+    @classmethod
+    def ndim_supp(cls, rng, weights, *components):
+        # We already checked that all components have the same dimensionality
```

Comment:
I think that the signature of this method should be `ndim_supp(cls, weights, *components)` since an `rng` doesn't seem to be provided [here](https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/distribution.py#L587)

Reply:
Good catch, I think you're right!